[v1.0.0b] - QueuePool limit reached when website idling in browser.

### First Check

- [X] This is not a feature request
- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.

### What is the issue you are experiencing?

I left the website idling in a browser (current Firefox for Windows 10), I'm not sure how long, maybe a day or so. When I came back to it, I was unable to log in. I checked the logs on the API container and I saw the following:
```
192.168.192.4:35516 - "POST /api/auth/token HTTP/1.1" 500
[2022-10-07 01:03:43 -0300] [164089] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 398, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 24, in __call__
    await responder(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 43, in __call__
    await self.app(scope, receive, self.send_with_gzip)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/exceptions.py", line 93, in __call__
    raise exc
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 266, in handle
    await self.app(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 65, in app
    response = await func(request)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 227, in app
    raw_response = await run_endpoint_function(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/app/mealie/routes/auth/auth.py", line 56, in get_token
    user = authenticate_user(session, email, password)  # type: ignore
  File "/app/mealie/core/security/security.py", line 82, in authenticate_user
    user = db.users.get_one(email, "email", any_case=True)
  File "/app/mealie/repos/repository_generic.py", line 135, in get_one
    result = q.one_or_none()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2845, in one_or_none
    return self._iter().one_or_none()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2903, in _iter
    result = self.session.execute(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1711, in execute
    conn = self._connection_for_bind(bind)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1552, in _connection_for_bind
    return self._transaction._connection_for_bind(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 747, in _connection_for_bind
    conn = bind.connect()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3234, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
    else engine.raw_connection()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3313, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3280, in _wrap_pool_connect
    return fn()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 868, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 476, in checkout
    rec = pool._do_get()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 134, in _do_get
    raise exc.TimeoutError(sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/14/3o7r)
127.0.0.1:34920 - "GET /docs HTTP/1.1" 200
192.168.192.4:35556 - "POST /api/auth/token HTTP/1.1" 500
[2022-10-07 01:04:02 -0300] [164089] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 398, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 24, in __call__
    await responder(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 43, in __call__
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/exceptions.py", line 93, in __call__
    raise exc
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 266, in handle
    await self.app(scope, receive, send)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/routing.py", line 65, in app
    response = await func(request)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 227, in app
    raw_response = await run_endpoint_function(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/app/mealie/routes/auth/auth.py", line 56, in get_token
    user = authenticate_user(session, email, password)  # type: ignore
  File "/app/mealie/core/security/security.py", line 82, in authenticate_user
    user = db.users.get_one(email, "email", any_case=True)
  File "/app/mealie/repos/repository_generic.py", line 135, in get_one
    result = q.one_or_none()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2845, in one_or_none
    return self._iter().one_or_none()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2903, in _iter
    result = self.session.execute(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1711, in execute
    conn = self._connection_for_bind(bind)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1552, in _connection_for_bind
    return self._transaction._connection_for_bind(
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 747, in _connection_for_bind
    conn = bind.connect()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3234, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 96, in __init__
    else engine.raw_connection()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3313, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 3280, in _wrap_pool_connect
    return fn()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 310, in connect
    return _ConnectionFairy._checkout(self)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 868, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/base.py", line 476, in checkout
    rec = pool._do_get()
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/sqlalchemy/pool/impl.py", line 134, in _do_get
    raise exc.TimeoutError(
sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/14/3o7r)
```

When I went back a bit in the logs to see what was happening when it was actually idling, I saw this:
```
127.0.0.1:56994 - "GET /docs HTTP/1.1" 200
127.0.0.1:57022 - "GET /docs HTTP/1.1" 200
127.0.0.1:57118 - "GET /docs HTTP/1.1" 200
127.0.0.1:57170 - "GET /docs HTTP/1.1" 200
127.0.0.1:57230 - "GET /docs HTTP/1.1" 200
INFO: 07-Oct-22 00:02:26        Running minutely callbacks
127.0.0.1:57254 - "GET /docs HTTP/1.1" 200
ERROR: 07-Oct-22 00:02:56       Error in scheduled task func='post_group_webhooks': exception='QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/14/3o7r)'
127.0.0.1:57310 - "GET /docs HTTP/1.1" 200
127.0.0.1:57344 - "GET /docs HTTP/1.1" 200
127.0.0.1:57396 - "GET /docs HTTP/1.1" 200
```
I had to 'down' and 'up' the whole docker-compose stack to get it running again.

### Deployment

Docker (Linux)

### Deployment Details

I'm using the docker-compose example from this repo.

--- a/mealie/db/db_setup.py
+++ b/mealie/db/db_setup.py
@@ -31,22 +31,31 @@ SessionLocal, engine = sql_global_init(settings.DB_URL)  # type: ignore
 
 
 @contextmanager
-def with_session() -> Session:
+def session_context() -> Session:
+    """
+    session_context() provides a managed session to the database that is automatically
+    closed when the context is exited. This is the preferred method of accessing the
+    database.
+
+    Note: use `generate_session` when using the `Depends` function from FastAPI
+    """
     global SessionLocal
     sess = SessionLocal()
-
     try:
         yield sess
     finally:
         sess.close()
 
 
-def create_session() -> Session:
-    global SessionLocal
-    return SessionLocal()
-
-
 def generate_session() -> Generator[Session, None, None]:
+    """
+    WARNING: This function should _only_ be called when used with
+    using the `Depends` function from FastAPI. This function will leak
+    sessions if used outside of the context of a request.
+
+    Use `with_session` instead. That function will allow you to use the
+    session within a context manager
+    """
     global SessionLocal
     db = SessionLocal()
     try:

--- a/mealie/db/init_db.py
+++ b/mealie/db/init_db.py
@@ -9,7 +9,7 @@ from alembic.config import Config
 from alembic.runtime import migration
 from mealie.core import root_logger
 from mealie.core.config import get_app_settings
-from mealie.db.db_setup import create_session
+from mealie.db.db_setup import session_context
 from mealie.db.fixes.fix_slug_foods import fix_slug_food_names
 from mealie.repos.all_repositories import get_repositories
 from mealie.repos.repository_factory import AllRepositories
@@ -67,41 +67,40 @@ def connect(session: orm.Session) -> bool:
 
 
 def main():
-    session = create_session()
-
     # Wait for database to connect
     max_retry = 10
     wait_seconds = 1
 
-    while True:
-        if connect(session):
-            logger.info("Database connection established.")
-            break
+    with session_context() as session:
+        while True:
+            if connect(session):
+                logger.info("Database connection established.")
+                break
 
-        logger.error(f"Database connection failed. Retrying in {wait_seconds} seconds...")
-        max_retry -= 1
+            logger.error(f"Database connection failed. Retrying in {wait_seconds} seconds...")
+            max_retry -= 1
 
-        sleep(wait_seconds)
+            sleep(wait_seconds)
 
-        if max_retry == 0:
-            raise ConnectionError("Database connection failed - exiting application.")
+            if max_retry == 0:
+                raise ConnectionError("Database connection failed - exiting application.")
 
-    alembic_cfg = Config(str(PROJECT_DIR / "alembic.ini"))
-    if db_is_at_head(alembic_cfg):
-        logger.info("Migration not needed.")
-    else:
-        logger.info("Migration needed. Performing migration...")
-        command.upgrade(alembic_cfg, "head")
+        alembic_cfg = Config(str(PROJECT_DIR / "alembic.ini"))
+        if db_is_at_head(alembic_cfg):
+            logger.info("Migration not needed.")
+        else:
+            logger.info("Migration needed. Performing migration...")
+            command.upgrade(alembic_cfg, "head")
 
-    db = get_repositories(session)
+        db = get_repositories(session)
 
-    if db.users.get_all():
-        logger.info("Database exists")
-    else:
-        logger.info("Database contains no users, initializing...")
-        init_db(db)
+        if db.users.get_all():
+            logger.info("Database exists")
+        else:
+            logger.info("Database contains no users, initializing...")
+            init_db(db)
 
-    safe_try(lambda: fix_slug_food_names(db))
+        safe_try(lambda: fix_slug_food_names(db))
 
 
 if __name__ == "__main__":

--- a/mealie/scripts/reset_locked_users.py
+++ b/mealie/scripts/reset_locked_users.py
@@ -1,5 +1,5 @@
 from mealie.core import root_logger
-from mealie.db.db_setup import with_session
+from mealie.db.db_setup import session_context
 from mealie.repos.repository_factory import AllRepositories
 from mealie.services.user_services.user_service import UserService
 
@@ -13,7 +13,7 @@ def main():
 
     logger = root_logger.get_logger()
 
-    with with_session() as session:
+    with session_context() as session:
         repos = AllRepositories(session)
         user_service = UserService(repos)
 

--- a/mealie/services/event_bus_service/event_bus_listeners.py
+++ b/mealie/services/event_bus_service/event_bus_listeners.py
@@ -10,7 +10,7 @@ from fastapi.encoders import jsonable_encoder
 from pydantic import UUID4
 from sqlalchemy.orm.session import Session
 
-from mealie.db.db_setup import with_session
+from mealie.db.db_setup import session_context
 from mealie.db.models.group.webhooks import GroupWebhooksModel
 from mealie.repos.all_repositories import get_repositories
 from mealie.repos.repository_factory import AllRepositories
@@ -52,7 +52,7 @@ class EventListenerBase(ABC):
         run as a scheduled task, the session is not provided and must be created.
         """
         if self.session is None:
-            with with_session() as session:
+            with session_context() as session:
                 self.session = session
                 yield
 

--- a/mealie/services/event_bus_service/event_bus_listeners.py
+++ b/mealie/services/event_bus_service/event_bus_listeners.py
@@ -1,4 +1,7 @@
+import contextlib
 import json
+from abc import ABC, abstractmethod
+from collections.abc import Generator
 from datetime import datetime, timezone
 from typing import cast
 from urllib.parse import parse_qs, urlencode, urlsplit, urlunsplit
@@ -7,6 +10,7 @@ from fastapi.encoders import jsonable_encoder
 from pydantic import UUID4
 from sqlalchemy.orm.session import Session
 
+from mealie.db.db_setup import with_session
 from mealie.db.models.group.webhooks import GroupWebhooksModel
 from mealie.repos.all_repositories import get_repositories
 from mealie.repos.repository_factory import AllRepositories
@@ -18,19 +22,42 @@ from .event_types import Event, EventDocumentType, EventTypes, EventWebhookData
 from .publisher import ApprisePublisher, PublisherLike, WebhookPublisher
 
 
-class EventListenerBase:
+class EventListenerBase(ABC):
+    session: Session | None
+
     def __init__(self, session: Session, group_id: UUID4, publisher: PublisherLike) -> None:
         self.session = session
         self.group_id = group_id
         self.publisher = publisher
 
+    @abstractmethod
     def get_subscribers(self, event: Event) -> list:
         """Get a list of all subscribers to this event"""
         ...
 
+    @abstractmethod
     def publish_to_subscribers(self, event: Event, subscribers: list) -> None:
         """Publishes the event to all subscribers"""
         ...
+
+    @contextlib.contextmanager
+    def ensure_session(self) -> Generator[None, None, None]:
+        """
+        ensure_session ensures that a session is available for the caller by checking if a session
+        was provided during construction, and if not, creating a new session with the `with_session`
+        function and closing it when the context manager exits.
+
+        This is _required_ when working with sessions inside an event bus listener where the listener
+        may be constructed during a request where the session is provided by the request, but the when
+        run as a scheduled task, the session is not provided and must be created.
+        """
+        if self.session is None:
+            with with_session() as session:
+                self.session = session
+                yield
+
+        else:
+            yield
 
 
 class AppriseEventListener(EventListenerBase):
@@ -38,14 +65,15 @@ class AppriseEventListener(EventListenerBase):
         super().__init__(session, group_id, ApprisePublisher())
 
     def get_subscribers(self, event: Event) -> list[str]:
-        repos = AllRepositories(self.session)
+        with self.ensure_session():
+            repos = AllRepositories(self.session)
 
-        notifiers: list[GroupEventNotifierPrivate] = repos.group_event_notifier.by_group(  # type: ignore
-            self.group_id
-        ).multi_query({"enabled": True}, override_schema=GroupEventNotifierPrivate)
+            notifiers: list[GroupEventNotifierPrivate] = repos.group_event_notifier.by_group(  # type: ignore
+                self.group_id
+            ).multi_query({"enabled": True}, override_schema=GroupEventNotifierPrivate)
 
-        urls = [notifier.apprise_url for notifier in notifiers if getattr(notifier.options, event.event_type.name)]
-        urls = AppriseEventListener.update_urls_with_event_data(urls, event)
+            urls = [notifier.apprise_url for notifier in notifiers if getattr(notifier.options, event.event_type.name)]
+            urls = AppriseEventListener.update_urls_with_event_data(urls, event)
 
         return urls
 
@@ -120,12 +148,13 @@ class WebhookEventListener(EventListenerBase):
 
     def get_scheduled_webhooks(self, start_dt: datetime, end_dt: datetime) -> list[ReadWebhook]:
         """Fetches all scheduled webhooks from the database"""
-        return (
-            self.session.query(GroupWebhooksModel)
-            .where(
-                GroupWebhooksModel.enabled == True,  # noqa: E712 - required for SQLAlchemy comparison
-                GroupWebhooksModel.scheduled_time > start_dt.astimezone(timezone.utc).time(),
-                GroupWebhooksModel.scheduled_time <= end_dt.astimezone(timezone.utc).time(),
+        with self.ensure_session():
+            return (
+                self.session.query(GroupWebhooksModel)
+                .where(
+                    GroupWebhooksModel.enabled == True,  # noqa: E712 - required for SQLAlchemy comparison
+                    GroupWebhooksModel.scheduled_time > start_dt.astimezone(timezone.utc).time(),
+                    GroupWebhooksModel.scheduled_time <= end_dt.astimezone(timezone.utc).time(),
+                )
+                .all()
             )
-            .all()
-        )

--- a/mealie/services/event_bus_service/event_bus_service.py
+++ b/mealie/services/event_bus_service/event_bus_service.py
@@ -40,12 +40,13 @@ class EventSource:
 
 
 class EventBusService:
+    bg: BackgroundTasks | None
+    session: Session | None
+    group_id: UUID4 | None
+
     def __init__(
         self, bg: Optional[BackgroundTasks] = None, session: Optional[Session] = None, group_id: UUID4 | None = None
     ) -> None:
-        if not session:
-            session = next(generate_session())
-
         self.bg = bg
         self.session = session
         self.group_id = group_id

--- a/mealie/services/scheduler/tasks/post_webhooks.py
+++ b/mealie/services/scheduler/tasks/post_webhooks.py
@@ -3,7 +3,7 @@ from typing import Optional
 
 from pydantic import UUID4
 
-from mealie.db.db_setup import with_session
+from mealie.db.db_setup import session_context
 from mealie.repos.all_repositories import get_repositories
 from mealie.schema.response.pagination import PaginationQuery
 from mealie.services.event_bus_service.event_bus_service import EventBusService
@@ -32,7 +32,7 @@ def post_group_webhooks(start_dt: Optional[datetime] = None, group_id: Optional[
     if group_id is None:
         # publish the webhook event to each group's event bus
 
-        with with_session() as session:
+        with session_context() as session:
             repos = get_repositories(session)
             groups_data = repos.groups.page_all(PaginationQuery(page=1, per_page=-1))
             group_ids = [group.id for group in groups_data.items]

--- a/mealie/services/scheduler/tasks/post_webhooks.py
+++ b/mealie/services/scheduler/tasks/post_webhooks.py
@@ -3,7 +3,7 @@ from typing import Optional
 
 from pydantic import UUID4
 
-from mealie.db.db_setup import create_session
+from mealie.db.db_setup import with_session
 from mealie.repos.all_repositories import get_repositories
 from mealie.schema.response.pagination import PaginationQuery
 from mealie.services.event_bus_service.event_bus_service import EventBusService
@@ -31,10 +31,11 @@ def post_group_webhooks(start_dt: Optional[datetime] = None, group_id: Optional[
 
     if group_id is None:
         # publish the webhook event to each group's event bus
-        session = create_session()
-        repos = get_repositories(session)
-        groups_data = repos.groups.page_all(PaginationQuery(page=1, per_page=-1))
-        group_ids = [group.id for group in groups_data.items]
+
+        with with_session() as session:
+            repos = get_repositories(session)
+            groups_data = repos.groups.page_all(PaginationQuery(page=1, per_page=-1))
+            group_ids = [group.id for group in groups_data.items]
 
     else:
         group_ids = [group_id]

--- a/mealie/services/scheduler/tasks/purge_group_exports.py
+++ b/mealie/services/scheduler/tasks/purge_group_exports.py
@@ -3,7 +3,7 @@ from pathlib import Path
 
 from mealie.core import root_logger
 from mealie.core.config import get_app_dirs
-from mealie.db.db_setup import create_session
+from mealie.db.db_setup import session_context
 from mealie.db.models.group.exports import GroupDataExportsModel
 
 ONE_DAY_AS_MINUTES = 1440
@@ -15,20 +15,19 @@ def purge_group_data_exports(max_minutes_old=ONE_DAY_AS_MINUTES):
 
     logger.info("purging group data exports")
     limit = datetime.datetime.now() - datetime.timedelta(minutes=max_minutes_old)
-    session = create_session()
 
-    results = session.query(GroupDataExportsModel).filter(GroupDataExportsModel.expires <= limit)
+    with session_context() as session:
+        results = session.query(GroupDataExportsModel).filter(GroupDataExportsModel.expires <= limit)
 
-    total_removed = 0
-    for result in results:
-        session.delete(result)
-        Path(result.path).unlink(missing_ok=True)
-        total_removed += 1
+        total_removed = 0
+        for result in results:
+            session.delete(result)
+            Path(result.path).unlink(missing_ok=True)
+            total_removed += 1
 
-    session.commit()
-    session.close()
+        session.commit()
 
-    logger.info(f"finished purging group data exports. {total_removed} exports removed from group data")
+        logger.info(f"finished purging group data exports. {total_removed} exports removed from group data")
 
 
 def purge_excess_files() -> None:

--- a/mealie/services/scheduler/tasks/purge_password_reset.py
+++ b/mealie/services/scheduler/tasks/purge_password_reset.py
@@ -1,7 +1,7 @@
 import datetime
 
 from mealie.core import root_logger
-from mealie.db.db_setup import create_session
+from mealie.db.db_setup import session_context
 from mealie.db.models.users.password_reset import PasswordResetModel
 
 logger = root_logger.get_logger()
@@ -13,8 +13,9 @@ def purge_password_reset_tokens():
     """Purges all events after x days"""
     logger.info("purging password reset tokens")
     limit = datetime.datetime.now() - datetime.timedelta(days=MAX_DAYS_OLD)
-    session = create_session()
-    session.query(PasswordResetModel).filter(PasswordResetModel.created_at <= limit).delete()
-    session.commit()
-    session.close()
-    logger.info("password reset tokens purges")
+
+    with session_context() as session:
+        session.query(PasswordResetModel).filter(PasswordResetModel.created_at <= limit).delete()
+        session.commit()
+        session.close()
+        logger.info("password reset tokens purges")

--- a/mealie/services/scheduler/tasks/purge_registration.py
+++ b/mealie/services/scheduler/tasks/purge_registration.py
@@ -1,7 +1,7 @@
 import datetime
 
 from mealie.core import root_logger
-from mealie.db.db_setup import create_session
+from mealie.db.db_setup import session_context
 from mealie.db.models.group import GroupInviteToken
 
 logger = root_logger.get_logger()
@@ -13,8 +13,10 @@ def purge_group_registration():
     """Purges all events after x days"""
     logger.info("purging expired registration tokens")
     limit = datetime.datetime.now() - datetime.timedelta(days=MAX_DAYS_OLD)
-    session = create_session()
-    session.query(GroupInviteToken).filter(GroupInviteToken.created_at <= limit).delete()
-    session.commit()
-    session.close()
+
+    with session_context() as session:
+        session.query(GroupInviteToken).filter(GroupInviteToken.created_at <= limit).delete()
+        session.commit()
+        session.close()
+
     logger.info("registration token purged")

--- a/mealie/services/scheduler/tasks/reset_locked_users.py
+++ b/mealie/services/scheduler/tasks/reset_locked_users.py
@@ -1,5 +1,5 @@
 from mealie.core import root_logger
-from mealie.db.db_setup import with_session
+from mealie.db.db_setup import session_context
 from mealie.repos.repository_factory import AllRepositories
 from mealie.services.user_services.user_service import UserService
 
@@ -8,7 +8,7 @@ def locked_user_reset():
     logger = root_logger.get_logger()
     logger.info("resetting locked users")
 
-    with with_session() as session:
+    with session_context() as session:
         repos = AllRepositories(session)
         user_service = UserService(repos)
 

--- a/tests/integration_tests/user_tests/test_user_password_reset_service.py
+++ b/tests/integration_tests/user_tests/test_user_password_reset_service.py
@@ -3,7 +3,7 @@ import json
 import pytest
 from fastapi.testclient import TestClient
 
-from mealie.db.db_setup import create_session
+from mealie.db.db_setup import session_context
 from mealie.services.user_services.password_reset_service import PasswordResetService
 from tests.utils.factories import random_string
 from tests.utils.fixture_schemas import TestUser
@@ -31,10 +31,10 @@ def test_password_reset(api_client: TestClient, unique_user: TestUser, casing: s
                 cased_email += l.lower()
         cased_email
 
-    session = create_session()
-    service = PasswordResetService(session)
-    token = service.generate_reset_token(cased_email)
-    assert token is not None
+    with session_context() as session:
+        service = PasswordResetService(session)
+        token = service.generate_reset_token(cased_email)
+        assert token is not None
 
     new_password = random_string(15)
 
@@ -58,8 +58,6 @@ def test_password_reset(api_client: TestClient, unique_user: TestUser, casing: s
     new_token = json.loads(response.text).get("access_token")
     response = api_client.get(Routes.self, headers={"Authorization": f"Bearer {new_token}"})
     assert response.status_code == 200
-
-    session.close()
 
     # Test successful password reset
     response = api_client.post(Routes.base, json=payload)

--- a/tests/unit_tests/test_security.py
+++ b/tests/unit_tests/test_security.py
@@ -5,7 +5,7 @@ from pytest import MonkeyPatch
 from mealie.core import security
 from mealie.core.config import get_app_settings
 from mealie.core.dependencies import validate_file_token
-from mealie.db.db_setup import create_session
+from mealie.db.db_setup import session_context
 from tests.utils.factories import random_string
 
 
@@ -47,5 +47,8 @@ def test_ldap_authentication_mocked(monkeypatch: MonkeyPatch):
     monkeypatch.setattr(ldap, "initialize", ldap_initialize_mock)
 
     get_app_settings.cache_clear()
-    result = security.authenticate_user(create_session(), user, password)
+
+    with session_context() as session:
+        result = security.authenticate_user(session, user, password)
+
     assert result is False

